### PR TITLE
Add temporary interceptor support

### PR DESCRIPTION
**Pull request by michaelbausor**
Adds temporary support for an Interceptor-style interface in order to support logging of requests and responses via interceptors until interceptor support is available in gRPC.

This PR also adds a GapicInterceptor that can be used to implement common functionality such as metadata handling.

### CODE REVIEW

File: src/ApiCore/Transport/Grpc/UnaryInterceptor.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Google\ApiCore\Transport\Grpc;
+
+/**
+ * Temporary class to support an interceptor-like interface until gRPC interceptor support is
+ * available.
+ *
+ * @package Google\ApiCore\Transport\Grpc
+ */
+interface UnaryInterceptor
```
Comment (jdpedrie) [inline src/ApiCore/Transport/Grpc/UnaryInterceptor.php]: It's a fairly strong convention to put `Interface` in the name of interfaces. Could we call this `UnaryInterceptorInterface`, or is the naming constrained by what gRPC is going to implement?

File: src/ApiCore/Transport/Grpc/LoggingInterceptor.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     * @param array $options
+     * @param callable $continuation
+     * @return LoggingUnaryCall
+     */
+    public function interceptUnaryUnary(
+        $method,
+        $argument,
+        array $metadata,
+        array $options,
+        $continuation
```
[the remark this answers is not in the document]
Comment (michaelbausor) [reply src/ApiCore/Transport/Grpc/LoggingInterceptor.php]: Done

File: src/ApiCore/Transport/Grpc/UnaryInterceptor.php (hunk 1 above)
Comment (michaelbausor) [reply src/ApiCore/Transport/Grpc/UnaryInterceptor.php]: Updated to UnaryInterceptorInterface

File: src/ApiCore/Transport/GrpcTransport.php [hunk 3]
```diff
@@ -40,6 +40,8 @@
 use Google\ApiCore\GrpcSupportTrait;
 use Google\ApiCore\ServerStream;
 use Google\ApiCore\ServiceAddressTrait;
+use Google\ApiCore\Transport\Grpc\GapicInterceptorInterface;
```
Comment (dwsupplee) [inline src/ApiCore/Transport/GrpcTransport.php]: This looks to be unused.

File: src/ApiCore/Transport/GrpcTransport.php [hunk 4]
```diff
@@ -68,6 +88,8 @@ class GrpcTransport extends BaseStub implements TransportInterface
      *
      *    @type array $stubOpts Options used to construct the gRPC stub.
      *    @type Channel $channel Grpc channel to be used.
+     *    @type UnaryInterceptorInterface[] $interceptors *INTERNAL* Interceptor support, required until
```
Comment (dwsupplee) [inline src/ApiCore/Transport/GrpcTransport.php]: Should https://github.com/googleapis/gax-php/pull/182/files#diff-f84fac5ef33d7480e5a575b4b95f04dfR72 match this description as well? Also, WDYT about using the term "experimental" as we do elsewhere to indicate the type could potentially change?

File: src/ApiCore/Transport/GrpcTransport.php [hunk 5]
```diff
@@ -160,6 +183,45 @@ public function startServerStreamingCall(Call $call, array $options)
         );
     }
 
+    private function wrapExecuteWithInterceptor($execute, $interceptor)
```
Comment (dwsupplee) [inline src/ApiCore/Transport/GrpcTransport.php]: Just a small nit, could these include typehints? (callable, UnaryInterceptorInterface)

File: src/ApiCore/Transport/GrpcTransport.php (hunk 3 above)
Comment (michaelbausor) [reply src/ApiCore/Transport/GrpcTransport.php]: Fixed

File: src/ApiCore/Transport/GrpcTransport.php (hunk 4 above)
Comment (michaelbausor) [reply src/ApiCore/Transport/GrpcTransport.php]: Done. Agree that experimental is better.

File: src/ApiCore/Transport/GrpcTransport.php (hunk 5 above)
Comment (michaelbausor) [reply src/ApiCore/Transport/GrpcTransport.php]: Done

**Pull request merged**